flink/src/test/java/org/apache/iceberg/flink/sink/TestFlinkIcebergSinkV2.java

```diff
@@ -0,0 +1,330 @@
+        .enableCheckpointing(100L)
+        .setParallelism(parallelism)
+        .setMaxParallelism(parallelism);
+
+    tableLoader = new TestTableLoader(tableDir.getAbsolutePath());
+  }
+
+  private List<Snapshot> findValidSnapshots(Table table) {
+    List<Snapshot> validSnapshots = Lists.newArrayList();
+    for (Snapshot snapshot : table.snapshots()) {
+      if (snapshot.allManifests().stream().anyMatch(m -> snapshot.snapshotId() == m.snapshotId())) {
+        validSnapshots.add(snapshot);
+      }
+    }
+    return validSnapshots;
+  }
+
+  private void testChangeLogs(List<String> equalityFieldColumns,
+                              KeySelector<Row, Row> keySelector,
+                              List<List<Row>> elementsPerCheckpoint,
+                              List<List<Record>> expectedRecordsPerCheckpoint) throws Exception {
+    DataStream<Row> dataStream = env.addSource(new BoundedTestSource<>(elementsPerCheckpoint), ROW_TYPE_INFO);
+
+    // Shuffle by the equality key, so that different operations from the same key could be wrote in order when
+    // executing tasks in parallelism.
+    dataStream = dataStream.keyBy(keySelector);
+
+    FlinkSink.forRow(dataStream, SimpleDataUtil.FLINK_SCHEMA)
+        .tableLoader(tableLoader)
+        .tableSchema(SimpleDataUtil.FLINK_SCHEMA)
+        .writeParallelism(parallelism)
+        .equalityFieldColumns(equalityFieldColumns)
+        .build();
+
+    // Execute the program.
+    env.execute("Test Iceberg Change-Log DataStream.");
+
+    table.refresh();
+    List<Snapshot> snapshots = findValidSnapshots(table);
+    int expectedSnapshotNum = expectedRecordsPerCheckpoint.size();
+    Assert.assertEquals("Should have the expected snapshot number", expectedSnapshotNum, snapshots.size());
+
+    for (int i = 0; i < expectedSnapshotNum; i++) {
+      long snapshotId = snapshots.get(i).snapshotId();
+      List<Record> expectedRecords = expectedRecordsPerCheckpoint.get(i);
+      Assert.assertEquals("Should have the expected records for the checkpoint#" + i,
+          expectedRowSet(expectedRecords.toArray(new Record[0])), actualRowSet(snapshotId, "*"));
+    }
+  }
+
+  private Row row(String rowKind, int id, String data) {
+    Map<String, RowKind> mapping = ImmutableMap.of(
+        "+I", RowKind.INSERT,
+        "-D", RowKind.DELETE,
+        "-U", RowKind.UPDATE_BEFORE,
+        "+U", RowKind.UPDATE_AFTER);
+
+    RowKind kind = mapping.get(rowKind);
+    if (kind == null) {
+      throw new IllegalArgumentException("Unknown row kind: " + rowKind);
+    }
+
+    return Row.ofKind(kind, id, data);
+  }
+
+  private Record record(int id, String data) {
+    return SimpleDataUtil.createRecord(id, data);
+  }
+
+  @Test
+  public void testChangeLogOnIdKey() throws Exception {
+    List<String> equalityFieldIds = ImmutableList.of("id");
+    List<List<Row>> elementsPerCheckpoint = ImmutableList.of(
+        ImmutableList.of(
+            row("+I", 1, "aaa"),
```

Comment:
Minor: This makes it look like the row has an operation as its first column, but that doesn't align with the key selector below that uses `row.getField(0)` to get the ID. I think it would make tests easier to read if `row` passed the row kind at the end. That way the fields align.

I'm not sure if it is worth changing all of the rows. Up to you.

Reply:
The current way is correct because it will maintain `rowKind` in a separate field ( rather than in the shared `fields` array) , see [here](https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/types/Row.java#L72).